nixos/modules/services/continuous-integration/hercules-ci-agent/deploy-keys.nix

```diff
@@ -0,0 +1,73 @@
+{ config, lib, options, pkgs, ... }:
```

Comment:
We shouldn't read nixops config options inside nixpkgs IMHO. Can we just ask for a location to certain files containing secrets (with sane defaults), and ask the user however they seem fit?

Reply:
I'm not sure if I understand correctly, because it's not really reading as much as writing nixops configuration.
Are you suggesting to provide options such that

    services.hercules-ci-agent.nixops.clusterJoinToken.keyFile = ./cluster-join-token;
    services.hercules-ci-agent.nixops.binaryCachesJson.keyFile = ./binary-caches.json;

cause the appropriate NixOps configs to be set?

Or do you disapprove of the entire concept of setting NixOps options for the user? I do think this is a rather nice possibility with NixOS/NixOps to make things a bit easier to set up.

Reply:
I've made it opt-in and the options now serve the dual purpose of documenting the default location for these files.

Reply:
Reading nixops options from inside nixpkgs is really a mispattern - how can we ever refactor things without breaking it on the other side?

NixOS modules really should just provide options to configure services to file locations (with sane defaults), and don't contain any deployment-tool specific code.

https://github.com/NixOS/rfcs/pull/59 proposes a generic API for these things, but until there's progress on that front, I'd really suggest providing options like `services.hercules-ci-agent.clusterJoinTokenFile` and `services.hercules-ci-agent.binaryCachesJsonFile` - see the gitlab module for example.

Reply:
If your change the interface of nixops, doesn't that break all consumers and isn't that easier to fix in NixOS than countless user deployments?
To answer the question literally: you can refactor by adding compatibility logic that reads the `options` module parameter, where all options can be introspected, including those from nixops.

`services.hercules-ci-agent.clusterJoinTokenFile` etc is already available via `extraOptions` if you need something custom for some reason.

I find this "mispattern" position rather disappointing. Hercules CI users are expecting this high-level behavior from the out-of-tree module now. Do I really have to bother them to needlessly think about those details again? It seems like a step backwards.

https://github.com/NixOS/rfcs/pull/59 doesn't seem like a sufficiently mature solution just yet. I *might* be able to help the RFC a bit, but I don't think it should be a blocker.

Reply:
> I find this "mispattern" position rather disappointing. Hercules CI users are expecting this high-level behavior from the out-of-tree module now. Do I really have to bother them to needlessly think about those details again? It seems like a step backwards.

There's currently no in-tree module using `deployment.keys`, and "it worked that way in an out-of-tree module" isn't really a justification by itself IMHO. Often things need to be refactored when sent mainline, to be more consistent with how things are done there.

Having the module reading secrets from a specific path by default (which is documented in the description and release notes/migration guide) seems less of a problem than using an option that isn't even documented in `man configuration.nix` and fails `nixos-rebuild`, when setting it on the machine:

```
error: The option `deployment' defined in `/etc/nixos/configuration.nix' does not exist.
(use '--show-trace' to show detailed location information)
```

Also, relying on things provided by nixops, or reading stuff from inside nixops has proven to be super painful. (See https://github.com/NixOS/nixops/pull/1156 and linked issues for details).

I assume people currently using the out-of-tree module need to do some sort of migration anyways (at least stop importing the out-of-tree module), so doing some more cleanup on the way shouldn't hurt.

Reply:
>  fails nixos-rebuild, when setting it on the machine:
> ```
> error: The option `deployment' defined in `/etc/nixos/configuration.nix' does not exist.
> (use '--show-trace' to show detailed location information)
> ```

fyi that was taken care of.

> Also, relying on things provided by nixops, or reading stuff from inside nixops has proven to be super painful.

At least I could fix it in a single place for everyone. It's even possible to adapt `config` to the `options` provided as was shown. Nonetheless I've decided not to continue the pioneering, because even if successful the cost/benefit seems a bit off.